cou/apps/base.py

```diff
@@ -848,7 +845,8 @@ def _check_application_target(self, target: OpenStackRelease) -> None:
         if (
             self.current_os_release >= target
             and not self.can_upgrade_to
-            and self.apt_source_codename >= target
+            # consider apt_source_codename just when exist or not empty
+            and (self.apt_source_codename >= target if self.os_origin else True)
```

Comment:
I don't understand this bit: if `os_origin` is empty, `apt_source_codename` will follow the `distro` codepath and return some OpenStack release codename. Can you give me an example of a situation that makes this change necessary?

Reply:
Sure...
Imagine an app like Vault that does not have `source` or `openstack-origin` or a subordinate-charm. When asked the `apt_source_codename` this will return `Ussuri` that will never be bigger than the target and won't halt the upgrade plan generation.

To not happen that, I'm considering just when there is something is set to `os_origin`

Reply:
@gabrielcocenza ok... but then you're leaking details pertaining to a different class into the base. In fact, the implementation of this function in `subordinate.py` is different.
...this shows the limitations of our class structure. In theory you could use the same approach and override this function in AuxiliaryApplication, except that while vault does not have a source setting, ceph-fs and ceph-radosgw do...

At this point perhaps - and not that I particularly like this - remove the override in `SubordinateApplication._check_application_target()` and handle all variations directly in base (you'll have to handle two different types of log lines though)

@samuelallan72 we have an ordered list of release codenames, the comparison is not purely alphabetical

Reply:
I start thinking, that it is time to start using mixin class. We are doing to many hack's in base already.

Reply:
Good catch. I've removed the function from the subordinate one because now the base can deal for all class types. Thanks